web/src/components/deleteButton.tsx

```diff
@@ -0,0 +1,116 @@
+import { useState } from "react";
+import { useRouter } from "next/router";
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from "@/src/components/ui/popover";
+import { Button } from "@/src/components/ui/button";
+import { TrashIcon } from "lucide-react";
+import { usePostHog } from "posthog-js/react";
+import { useHasAccess } from "@/src/features/rbac/utils/checkAccess";
+import { type Scope } from "@/src/features/rbac/constants/roleAccessRights";
+import { api } from "@/src/utils/api";
+
+interface DeleteButtonProps {
+  itemId: string;
+  projectId: string;
+  isTableAction?: boolean;
+  scope: Scope;
+  invalidateFunc: () => void;
+  type: "trace" | "dataset";
+  redirectUrl?: string;
+}
+
+export function DeleteButton({
+  itemId,
+  projectId,
+  isTableAction = false,
+  scope,
+  invalidateFunc,
+  type,
+  redirectUrl,
+}: DeleteButtonProps) {
+  const [isDeleted, setIsDeleted] = useState(false);
+  const router = useRouter();
+  const posthog = usePostHog();
+
+  const hasAccess = useHasAccess({ projectId, scope: scope });
+  const traceMutation = api.traces.deleteMany.useMutation({
```

Comment:
With this abstraction, some domain logic lives in this component (e.g. traceMutation), and some in the caller (e.g. scope). if we go for this abstraction, i would also extract these mutations.